File: src/main/java/org/cimsbioko/fragment/SyncHistoryFragment.java
package org.cimsbioko.fragment;

import android.os.Bundle;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.androidplot.xy.BoundaryMode;
import com.androidplot.xy.LineAndPointFormatter;
import com.androidplot.xy.SimpleXYSeries;
import com.androidplot.xy.StepMode;
import com.androidplot.xy.XYGraphWidget;
import com.androidplot.xy.XYPlot;
import com.androidplot.xy.XYSeries;

import org.cimsbioko.R;
import org.cimsbioko.provider.DatabaseAdapter;

import java.text.FieldPosition;
import java.text.Format;
import java.text.ParsePosition;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

public class SyncHistoryFragment extends Fragment {

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        View graphView = inflater.inflate(R.layout.sync_history_fragment, container, false);

        XYPlot plot = graphView.findViewById(R.id.sync_history_plot);

        DatabaseAdapter db = DatabaseAdapter.getInstance(getActivity());

        XYSeries series = new SimpleXYSeries(SimpleXYSeries.ArrayFormat.XY_VALS_INTERLEAVED, "", db.getSyncResults());

        LineAndPointFormatter formatter = new LineAndPointFormatter();
        formatter.setFillPaint(null);
        formatter.getVertexPaint().setARGB(255, 0, 0, 0);
        plot.addSeries(series, formatter);

        final int SECONDS_IN_DAY = 86400;
        final float MAX_DIVISIONS = 15.0f;
        plot.getLegend().setVisible(false);
        plot.getDomainTitle().setVisible(false);
        plot.setPlotPadding(5, 5, 5, 0);

        double daysToStep = 1;
        if (series.size() > 1) {
            final int duration = series.getX(series.size() - 1).intValue() - series.getX(0).intValue();
            final int daysSpanned = duration / SECONDS_IN_DAY;
            daysToStep = Math.max(1.0, Math.ceil(daysSpanned / MAX_DIVISIONS));
        }
        final double secondsToStep = daysToStep * SECONDS_IN_DAY;
        plot.setDomainStep(StepMode.INCREMENT_BY_VAL, secondsToStep);

        if (series.size() > 0) {
            plot.setDomainLowerBoundary(getMidnightOfDay(series.getX(0).longValue()), BoundaryMode.FIXED);
            plot.setDomainUpperBoundary(getMidnightOfDay(series.getX(series.size()-1).longValue()) + (long)secondsToStep, BoundaryMode.FIXED);
        }

        XYGraphWidget graph = plot.getGraph();

        XYGraphWidget.LineLabelStyle bottomStyle = graph.getLineLabelStyle(XYGraphWidget.Edge.BOTTOM);
        bottomStyle.setFormat(new SyncHistoryTimeFormat());
        bottomStyle.setRotation(-75);

        return graphView;
    }

    private long getMidnightOfDay(long epochTime) {
        final int MILLIS_IN_SEC = 1000;
        Calendar startCal = Calendar.getInstance();
        startCal.setTimeInMillis(epochTime * MILLIS_IN_SEC);
        startCal.set(Calendar.HOUR_OF_DAY, 0);
        startCal.set(Calendar.MINUTE, 0);
        startCal.set(Calendar.SECOND, 0);
        startCal.set(Calendar.MILLISECOND, 0);
        return startCal.getTimeInMillis() / MILLIS_IN_SEC;
    }

    private static class SyncHistoryTimeFormat extends Format {

        private static final int MILLIS_IN_SEC = 1000;

        private SimpleDateFormat dateFormat = new SimpleDateFormat("MM/dd");

        @Override
        public StringBuffer format(Object obj, StringBuffer buffer, FieldPosition field) {
            long timestamp = ((Number) obj).longValue() * MILLIS_IN_SEC;
            Date date = new Date(timestamp);
            return dateFormat.format(date, buffer, field);
        }

        @Override
        public Object parseObject(String string, ParsePosition position) {
            return null;
        }
    }
}

File: src/main/kotlin/org/cimsbioko/fragment/SyncHistoryFragment.kt
package org.cimsbioko.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.androidplot.xy.*
import org.cimsbioko.databinding.SyncHistoryFragmentBinding
import org.cimsbioko.provider.DatabaseAdapter
import java.text.FieldPosition
import java.text.Format
import java.text.ParsePosition
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.ceil
import kotlin.math.max

private const val MILLIS_IN_SEC = 1000
private const val SECONDS_IN_DAY = 86400
private const val MAX_DIVISIONS = 15.0f

class SyncHistoryFragment : Fragment() {

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return SyncHistoryFragmentBinding.inflate(inflater, container, false).also {
            it.syncHistoryPlot.apply {
                SimpleXYSeries(SimpleXYSeries.ArrayFormat.XY_VALS_INTERLEAVED, "", *DatabaseAdapter.syncResults).also { series ->
                    addSeries(series, LineAndPointFormatter().apply {
                        fillPaint = null
                        vertexPaint.setARGB(255, 0, 0, 0)
                    })
                    legend.isVisible = false
                    domainTitle.isVisible = false
                    setPlotPadding(5f, 5f, 5f, 0f)
                    val secondsToStep = (if (series.size() > 1) {
                        val duration: Int = series.getX(series.size() - 1).toInt() - series.getX(0).toInt()
                        val daysSpanned: Int = duration / SECONDS_IN_DAY
                        max(1.0, ceil(daysSpanned / MAX_DIVISIONS.toDouble()))
                    } else 1.0) * SECONDS_IN_DAY
                    setDomainStep(StepMode.INCREMENT_BY_VAL, secondsToStep)
                    if (series.size() > 0) {
                        setDomainLowerBoundary(getMidnightOfDay(series.getX(0).toLong()), BoundaryMode.FIXED)
                        setDomainUpperBoundary(getMidnightOfDay(series.getX(series.size() - 1).toLong()) + secondsToStep.toLong(), BoundaryMode.FIXED)
                    }
                    graph.getLineLabelStyle(XYGraphWidget.Edge.BOTTOM).apply {
                        format = SyncHistoryTimeFormat()
                        rotation = -75f
                    }
                }
            }
        }.root
    }

    private fun getMidnightOfDay(epochTime: Long): Long =
            Calendar.getInstance()
                    .also { c ->
                        c.timeInMillis = epochTime * MILLIS_IN_SEC
                        c[Calendar.HOUR_OF_DAY] = 0
                        c[Calendar.MINUTE] = 0
                        c[Calendar.SECOND] = 0
                        c[Calendar.MILLISECOND] = 0
                    }.timeInMillis / MILLIS_IN_SEC
}


private class SyncHistoryTimeFormat : Format() {

    private val dateFormat = SimpleDateFormat("MM/dd")

    override fun format(obj: Any, buffer: StringBuffer, field: FieldPosition): StringBuffer {
        val timestamp: Long = (obj as Number).toLong() * MILLIS_IN_SEC
        val date = Date(timestamp)
        return dateFormat.format(date, buffer, field)
    }

    override fun parseObject(string: String, position: ParsePosition): Any? = null
}